app/assets/javascripts/image-upload-markdown.js

```diff
@@ -0,0 +1,60 @@
+// This is a modified version of https://github.com/stuartnelson3/blog/blob/master/public/js/script.js
+
+
+$(document).on('drop', '.js-markdown-image-upload', function(e) {
+
+  e.preventDefault();
+  var dt = e.originalEvent.dataTransfer;
+  for (var i = 0; i < dt.files.length; i++) {
+    var file = dt.files[i];
+    var textarea = e.currentTarget;
+    handleFileUpload(textarea, file);
+  }
+
+  function addImageMarkdownPlaceholder(e, filename) {
+    var startPos = e.selectionStart;
+    var endPos = e.selectionEnd;
+    e.value = e.value.substring(0, startPos)
+      + "![Uploading " + filename + "...]()\r\n"
+      + e.value.substring(endPos, e.value.length);
+  }
+
+  function replaceImageMarkdownPlaceholder(textarea, value, filename){
+    var re = new RegExp("!\\[Uploading " + filename + "\\.\\.\\.\\]\\(\\)");
+    textarea.value = textarea.value.replace(re, function($1) {
+      return $1.replace("![Uploading " + filename + "...]()", value);
+    });
+  }
+
+  function handleFileUpload(textarea, file) {
+    addImageMarkdownPlaceholder(textarea, file.name);
+
+    var fd = new FormData();
+    fd.append("format", "js");
+    fd.append("file", file);
+
+    // CSRF token
+    fd.append("authenticity_token", $('meta[name="csrf-token"]').attr("content"));
+
+    var request = new XMLHttpRequest();
+    request.open("POST", "/uploads");
+
+    request.onloadend = function(filename) {
+      // Form a closure around the filename.
```

Comment:
We need this because multiple file uploads can happen asynchronously. There is a more detailed explanation [on the blog](http://www.stuart-nelson.co.uk/github-style-drag-n-drop-image-uploads) (cmd +f "closure")